src/Illuminate/Queue/SerializesModels.php

```diff
@@ -11,67 +10,105 @@ trait SerializesModels
+            }
+
+            $property->setValue($this, $this->getRestoredPropertyValue(
+                $this->getPropertyValue($property)
+            ));
+        }
+    }
+
+    /**
+     * Prepare the instance values for serialization.
      *
      * @return array
      */
-    public function __sleep()
+    public function __serialize()
     {
+        $values = [];
+
         $properties = (new ReflectionClass($this))->getProperties();
 
         foreach ($properties as $property) {
-            if ($property->getName() === 'modelIdentifiers') {
+            if ($property->isStatic()) {
                 continue;
             }
 
-            $serializedValue = $this->getSerializedPropertyValue(
-                $value = $this->getPropertyValue($property)
-            );
+            $name = $property->getName();
 
-            if ($serializedValue instanceof ModelIdentifier) {
-                $this->modelIdentifiers[$property->getName()] = $serializedValue;
-
-                // Empty instance of the model or collection to support typed properties...
-                $property->setValue($this, new $value);
-            } else {
-                $property->setValue($this, $value);
+            if ($property->isPrivate()) {
+                $name = "\0A\0{$name}";
+            } elseif ($property->isProtected()) {
+                $name = "\0*\0{$name}";
             }
+
+            $values[$name] = $this->getSerializedPropertyValue($this->getPropertyValue($property));
         }
 
-        return array_values(array_filter(array_map(function ($p) {
-            return $p->isStatic() ? null : $p->getName();
-        }, $properties)));
+        return $values;
     }
 
     /**
      * Restore the model after serialization.
      *
-     * @return void
+     * @param  array  $values
+     * @return array
      */
-    public function __wakeup()
+    public function __unserialize(array $values)
     {
-        foreach ((new ReflectionClass($this))->getProperties() as $property) {
-            if ($property->isStatic() || $property->getName() === 'modelIdentifiers') {
+        $properties = (new ReflectionClass($this))->getProperties();
+
+        foreach ($properties as $property) {
+            if ($property->isStatic()) {
                 continue;
             }
 
-            if (isset($this->modelIdentifiers[$property->getName()])) {
-                $value = $this->modelIdentifiers[$property->getName()];
-            } else {
-                $value = $this->getPropertyValue($property);
+            $name = $property->getName();
+
+            if ($property->isPrivate()) {
+                $name = "\0A\0{$name}";
+            } elseif ($property->isProtected()) {
+                $name = "\0*\0{$name}";
             }
 
-            $property->setValue($this, $this->getRestoredPropertyValue(
-                $value
-            ));
+            if (! array_key_exists($name, $values)) {
+                continue;
+            }
```

Comment:
This part will unfortunately break jobs which are already queued. If any app is upgraded with these changes those jobs won't be able to unserialize anymore.

Reply:
Why? 
In previous implementation anything won't be able unserialize anymore.

Reply:
PHP <7.4 does not use __serialize. If you will upgrade only Laravel php still used __slee/__wakeup.

Reply:
Because any already queued job will still have their model identifiers set on the properties. If you upgrade to php 7.4 then your queued jobs will break while with the current implementation you can upgrade to php 7.4 smoothly. 

We can do the unserialize thing in a year or two perhaps when we know most people have transitioned to php 7.4

Reply:
... then php pass all properties to __unserialize as array.

Reply:
I've added serialization structure test.

Reply:
Ah thanks for that. If @taylorotwell is okay with it then this PR is fine by me.